Simplify pass on arithmetic expressions for integers.

zhxchen17 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61444 [jit] Arithmetic simplification for integers.**

Summary:
Add a mini pass to merge arithmetic nodes like (((x - 1) + 2) * 1) - 1.
Issue #60913

Test Plan:
python test/test_jit.py TestPeephole.test_peephole_arith

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29630614](https://our.internmc.facebook.com/intern/diff/D29630614)

zhxchen17 committed:
[jit] Arithmetic simplification for integers.

Summary:
Add a mini pass to merge arithmetic nodes like (((x - 1) + 2) * 1) - 1.
Issue #60913

Test Plan:
python test/test_jit.py TestPeephole.test_peephole_arith

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29630614).

zhxchen17 committed:
Update on "[jit] Arithmetic simplification for integers."


Summary:
Add a mini pass to merge arithmetic nodes like (((x - 1) + 2) * 1) - 1.
Issue #60913

Test Plan:
python test/test_jit.py TestPeephole.test_peephole_arith

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29630614](https://our.internmc.facebook.com/intern/diff/D29630614)

[ghstack-poisoned]

zhxchen17 committed:
Update on "[jit] Arithmetic simplification for integers."


Summary:
Add a mini pass to merge arithmetic nodes like (((x - 1) + 2) * 1) - 1.
Issue #60913

Test Plan:
python test/test_jit.py TestPeephole.test_peephole_arith

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29630614](https://our.internmc.facebook.com/intern/diff/D29630614)

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29630614).

zhxchen17 committed:
Update on "[jit] Arithmetic simplification for integers."


Summary:
Add a mini pass to merge arithmetic nodes like (((x - 1) + 2) * 1) - 1.
Issue #60913

Test Plan:
python test/test_jit.py TestPeephole.test_peephole_arith

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29630614](https://our.internmc.facebook.com/intern/diff/D29630614)

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29630614).

zhxchen17 committed:
Update on "[jit] Arithmetic simplification for integers."


Summary:
Add a mini pass to merge arithmetic nodes like (((x - 1) + 2) * 1) - 1.
Issue #60913

Test Plan:
python test/test_jit.py TestPeephole.test_peephole_arith

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29630614](https://our.internmc.facebook.com/intern/diff/D29630614)

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29630614).

eellison commented on torch/csrc/jit/passes/peephole_non_tensor.cpp:
```diff
@@ -7,6 +7,131 @@
+    }
+  }
+
+  return getConstantInt(node, 1);
```
-> `constant_as<int64_t>`

eellison commented on torch/csrc/jit/passes/peephole_non_tensor.cpp:
```diff
@@ -7,6 +7,131 @@
+
+  const auto uses = node.output()->uses();
+  for (const auto& u : uses) {
+    if (tryReduceArith(*u.user)) {
```
Im a little confused by the flow here. We iterate through every node in the graph below in `PeepholeOptimizeNonTensorImpl::optimizeBlock`. That invokes, `tryReduceArith` which makes sense. Now, when we find the node, we also look forward into its future uses which we will be calling in `optimizeBlock` anyway. Is there a reason for this ?

eellison commented on torch/csrc/jit/passes/peephole_non_tensor.cpp:
```diff
@@ -7,6 +7,131 @@
+
+/**
+ * Merge an add/sub node with its users. If there exists a mul/floordiv node by
+ * 1, we firstly run tryReduceMulAndDiv() to bring all subsequent IR nodes as
```
WDYM by `ReduceMulAndDiv() to bring all subsequent IR nodes as direct IR user node`, not sure i follow

eellison commented on torch/csrc/jit/passes/peephole_non_tensor.cpp:
```diff
@@ -7,6 +7,131 @@
+    return true;
+  }
+
+  std::vector<Node*> mulAndDivToRemove;
```
This is another case where we are iterating forward looking through uses. As far as i understand, whenever we are looking at a node we are looking at all of its uses for cleanup, and then those nodes are looking in their uses for clean up, etc. 

If you have 

`add -> sub -> mul -> div`

add will recurse into sub->mul->div, then sub will recurse into mul->div, and mul will recurse in div. If we have a chain of n uses, and each use we iterate into future uses, we're doing O(n^2) work. 

I think it would be simpler and more efficient to just look at a node and see if it can be normalized wrt/ its inputs. When we look at sub, check if it can be normalized with add, when we look at mul we check if it can be normalized with sub, etc.

eellison commented on torch/csrc/jit/passes/peephole_non_tensor.cpp:
```diff
@@ -7,6 +7,131 @@
+    if (tryReduceMulAndDiv(*n)) {
+      n->destroy();
+    } else {
+      return false;
```
we dont want to not iterate through the other `mulAndDivToRemove` nodes because one failed right ?

eellison reviewed (commented):
looking good!! main comment is to see if we can iterate more simply, and then i had a couple small comments mostly about right apis to use

zhxchen17 commented:
update: after an offline discussion with @eellison , we clarified that it's better to just do "relaxation" style transformation by replacing backward the inputs of each arithmetic node instead of replacing forward the uses of an arithmetic node, so that the code is easier to maintain and the time complexity is linear (because we don't need to iterate users of a node), and it's okay in the case of doing extra work without destroying any node because the final goal is to simplify the dependency for shape propagation.
will update.

zhxchen17 committed:
Update on "[jit] Arithmetic simplification for integers."


Summary:
Add a mini pass to merge arithmetic nodes like (((x - 1) + 2) * 1) - 1.
Issue #60913

Test Plan:
python test/test_jit.py TestPeephole.test_peephole_arith

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29630614](https://our.internmc.facebook.com/intern/diff/D29630614)

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29630614).

zhxchen17 committed:
Update on "[jit] Arithmetic simplification for integers."


Summary:
Add a mini pass to merge arithmetic nodes like (((x - 1) + 2) * 1) - 1.
Issue #60913

Test Plan:
python test/test_jit.py TestPeephole.test_peephole_arith

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29630614](https://our.internmc.facebook.com/intern/diff/D29630614)

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29630614).

eellison reviewed (approved):
nice!! looks great!!

zhxchen17 committed:
Update on "[jit] Arithmetic simplification for integers."


Summary:
Add a mini pass to merge arithmetic nodes like (((x - 1) + 2) * 1) - 1.
Issue #60913

Test Plan:
python test/test_jit.py TestPeephole.test_peephole_arith

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29630614](https://our.internmc.facebook.com/intern/diff/D29630614)

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29630614).
